toqito/matrix_props/tests/test_is_k_incoherent.py

```diff
@@ -0,0 +1,228 @@
+        "toqito.matrix_props.is_k_incoherent", lambda Y, k_val, tol=1e-15: False if k_val == 2 else False
+    )
+
+    # Force the SDP branch by monkeypatching the solver to return 0.0 so that 1 - min(0.0, 1) equals 1.
+    def fake_solve(self, *args, **kwargs):
+        return 0.0
+
+    monkeypatch.setattr(cp.Problem, "solve", fake_solve)
+    assert is_k_incoherent(X, 3) is True
+
+
+@pytest.mark.parametrize(
+    "mat, k, expected",
+    [
+        # Explicitly constructed 3x3 matrix requiring SDP check, known from reference:
+        (np.array([[0.5, -0.5, 0.0], [-0.5, 0.5, 0.0], [0.0, 0.0, 0.0]]), 2, True),
+    ],
+)
+def test_sdp_branch(mat, k, expected):
+    """Explicitly test the SDP branch for k-incoherence."""
+    res = is_k_incoherent(mat, k)
+    assert res is True
+
+
+def test_non_square():
+    """Ensure non-square input is flagged."""
+    # Construct a non-square matrix.
+    X = np.array([[1, 2, 3], [4, 5, 6]])
+
+    with pytest.raises(ValueError, match="Input matrix must be square."):
+        is_k_incoherent(X, 2)
+
+
+def test_hierarchical_recursion(monkeypatch):
+    """Hierarchical recursion: for k >= 2 check incoherence for k-1."""
+    # Construct a 4x4 non-diagonal density matrix.
+    A = np.array(
+        [[0.36, 0.10, 0.05, 0.02], [0.08, 0.28, 0.04, 0.03], [0.05, 0.04, 0.20, 0.06], [0.02, 0.03, 0.06, 0.16]]
+    )
+    # A is non-diagonal and has trace close to 1.
+    A = A / np.trace(A)
+
+    # Save the original function.
+    orig_is_k_incoherent = is_k_incoherent
+
+    # Monkey-patch is_k_incoherent so that when called with k == 2 it returns True.
+    def fake_is_k_incoherent(mat, k, tol=1e-15):
+        if k == 2:
+            return True
+        else:
+            return orig_is_k_incoherent(mat, k, tol)
+
+    monkeypatch.setattr("toqito.matrix_props.is_k_incoherent", fake_is_k_incoherent)
+
+    # Now, when testing for k == 3, the recursive call (with k == 2) will return True,
+    # and the hierarchical recursion block will be taken.
+    result = is_k_incoherent(A, 3)
+    assert result is True
+
+
+def test_hierarchical_recursion_branch():
+    """Hierarchical recursion: for k =3 check incoherence for k-1."""
+    # Candidate 4x4 density matrix (non-diagonal, trace normalized to 1).
+    A2 = np.array(
+        [[0.35, 0.30, 0.00, 0.00], [0.30, 0.25, 0.00, 0.00], [0.00, 0.00, 0.25, 0.05], [0.00, 0.00, 0.05, 0.15]]
+    )
+    A2 = A2 / np.trace(A2)
+    k = 3
+
+    # Save the original recursive function.
+    orig_is_k_incoherent = is_k_incoherent.__globals__["is_k_incoherent"]
+
+    # Override is_k_incoherent so that when called with k==2 it returns True.
+    def fake_is_k_incoherent(mat, k, tol=1e-15):
+        if k == 2:
+            return True
+        else:
+            return orig_is_k_incoherent(mat, k, tol)
+
+    # Patch the global lookup so that recursive calls use fake_is_k_incoherent.
+    is_k_incoherent.__globals__["is_k_incoherent"] = fake_is_k_incoherent
```

Comment:
pretty neat. I did not know this was possible.

Reply:
I had to go looking for this solution, but yes, it was useful in this context!